### ISSUE TITLE
vm-overrides proposal

### ISSUE THREAD

**Issue by EdAyers** [leanprover-community/lean#9]
The purpose of VM-overrides is to be able to replace non-`meta` code with more efficient `meta` code when evaluating in the VM. Some use cases:

- Write a formally verified version of floating point numbers but use native floats in the VM.
   The result is that users can use floats in non-meta code without the performance penalty.
- Make finsets evaluatable using native rb trees?
- Efficient linear algebra, matrices.
- Make it clear in the Lean `init` code when a function is being overridden by Lean.

Non-goals:
- No guarantees that the meta and non-meta interpretations do the same thing. VM evaluation can still have bugs which the kernel doesn't see.
- Some typechecking will be done but no guarantees that the override and original declaration are compatible.

Here is a use case of VM-overrides as discussed on Zulip.

``` lean
-- These are implemented in C++ using native floats.
meta constant native.float : Type
meta constant native.float.add : native.float → native.float → native.float

inductive IEEE_float := ...

def IEEE_float.add := ...
-- plus other defininitions

meta constant native.float.to_IEEE : native.float → IEEE_float
meta constant native.float.of_IEEE : IEEE_float → native.float

-- No vm implementation needed
@[inline] meta def native.float.rec {C : Sort*}
  (H : IEEE_float → C) (f : native.float) : C :=
H f.to_IEEE

@[vm_override native.float] -- optional?
structure float := (to_IEEE : IEEE_float)

-- Override the core functions
attribute [vm_override native.float.rec] float.rec
attribute [vm_override native.float.of_IEEE] float.mk
attribute [vm_override native.float.to_IEEE] float.to_IEEE

@[vm_override native.float.add]
def float.add : float -> float - > float := \lam (mk x) (mk y), mk $ IEEE_float.add x y

-- No error because the equation compiler defines this function in terms of
-- float.rec and float.mk
def float.neg : float -> float  := \lam (mk x) , mk $ IEEE_float.neg x

-- This does not error because only composed of vm_overridden functions.
def float.double := \lam x , float.add x x

-- prints a big list `(float, native.float), (float.add, native.float.add), ...`
#print vm_override

-- prints a `native.float`
#print vm_override float.add

-- This _does_ error because the arities do not match up.
@[vm_override native.float.pow]
def float.square : float -> float := \lam (mk x), ...
```

**Comment by cipher1024** [comment]
Very nice! When you check the overrides, do you only check the arities?

**Comment by EdAyers** [comment]
I've got an arity checker working. I'm not going to try to get typechecking working.

**Comment by cipher1024** [comment]
I think we could try to do the type checking on the Lean side. I can give it a try once you're happy with the C++ side.

**Comment by cipher1024** [comment]
Here are more use cases I can think of:

* efficient implementation of qpf-based inductive / coinductive data types
* modelling the io monad using free monads and executing the model using the native `io` type

**Comment by EdAyers** [comment]
Ok checkout https://github.com/EdAyers/lean/tree/vm_override
and point Lean at https://github.com/EdAyers/lean/blob/vm_override/tests/lean/vm_override.lean

**Comment by cipher1024** [comment]
That looks really good! Is anything missing?

I just had the thought that we could use this to make a ffi. We could add an ffi attribute to constants and mention a .so file or a .dll file and specify the name of a C function. Then, with vm_override, we can use that C function to implement a Lean function

**Comment by khoek** [comment]
@cipher1024 That last thing sounds especially cool to me!

**Comment by cipher1024** [comment]
Awesome! It just came to me :) I didn't think we'd see a FFI in Lean 3 but now ... I think it's not so far fetched.

And I think nothing prevents us from creating the C files and .so files using Lean metaprogramming ...

**Comment by EdAyers** [comment]
Desired features:
- Typechecking or at least a sanity check
- Print the things that are overridden.
- Add a set_option that turns off vm_overrides.
- Better error handling, atm it is quite fragile if you don't use it right. I would at least like it to not need a server restart when it goes wrong.
- Foreign function stuff using attributes would be cool. I will maybe have a go at this at some point but not confident.

Bugs and problems:
- Lean creates different arity bytecode for `def x | (a,b) := z` vs `def x (p:_) := z` by automatically performing projections at the call site in the first case.  At best this triggers a counterintuitive arity missmatch and at worst causes a VM crash.
- I am not confident enough I understand the move semantics conventions in lean well enough to be sure I haven't introduced memory leaks etc.
- It needs more testing in general.

**Comment by gebner** [comment]
Merged in #48.
